sdks/python/apache_beam/transforms/util.py

```diff
@@ -161,3 +174,215 @@ def RemoveDuplicates(pcoll):  # pylint: disable=invalid-name
+          min(self._max_batch_size,
+              self._min_batch_size * self.MAX_GROWTH_FACTOR),
+          self._min_batch_size + 1))
+
+    # Linear regression for y = a + bx, where x is batch size and y is time.
+    xs, ys = zip(*self._data)
+    n = float(len(self._data))
+    xbar = sum(xs) / n
+    ybar = sum(ys) / n
+    b = (sum([(x - xbar) * (y - ybar) for x, y in self._data])
+         / sum([(x - xbar)**2 for x in xs]))
+    a = ybar - b * xbar
+
+    # Avoid nonsensical or division-by-zero errors below due to noise.
+    a = max(a, 1e-10)
+    b = max(b, 1e-20)
+
+    last_batch_size = self._data[-1][0]
+    cap = min(last_batch_size * self.MAX_GROWTH_FACTOR, self._max_batch_size)
+
+    if self._target_batch_duration:
+      # Solution to a + b*x = self._target_batch_duration.
+      cap = min(cap, (self._target_batch_duration - a) / b)
+
+    if self._target_batch_overhead:
+      # Solution to a / (a + b*x) = self._target_batch_overhead.
+      cap = min(cap, (a / b) * (1 / self._target_batch_overhead - 1))
+
+    # Avoid getting stuck at min_batch_size.
+    jitter = len(self._data) % 2
+    return int(max(self._min_batch_size + jitter, cap))
+
+
+class _GlobalWindowsBatchingDoFn(DoFn):
+  def __init__(self, batch_size_estimator):
+    self._batch_size_estimator = batch_size_estimator
+
+  def start_bundle(self):
+    self._batch = []
+    self._batch_size = self._batch_size_estimator.next_batch_size()
+    # The first emit often involves non-trivial setup.
+    self._batch_size_estimator.ignore_next_timing()
+
+  def process(self, element):
+    self._batch.append(element)
+    if len(self._batch) >= self._batch_size:
+      with self._batch_size_estimator.record_time(self._batch_size):
+        yield self._batch
+      self._batch = []
+      self._batch_size = self._batch_size_estimator.next_batch_size()
+
+  def finish_bundle(self):
+    if self._batch:
+      with self._batch_size_estimator.record_time(self._batch_size):
+        yield window.GlobalWindows.windowed_value(self._batch)
+      self._batch = None
+      self._batch_size = self._batch_size_estimator.next_batch_size()
+
+
+class _WindowAwareBatchingDoFn(DoFn):
+  def __init__(self, batch_size_estimator):
+    self._batch_size_estimator = batch_size_estimator
+
+  def start_bundle(self):
+    self._batches = collections.defaultdict(list)
+    self._batch_size = self._batch_size_estimator.next_batch_size()
+    # The first emit often involves non-trivial setup.
+    self._batch_size_estimator.ignore_next_timing()
+
+  def process(self, element, window=DoFn.WindowParam):
+    self._batches[window].append(element)
+    if len(self._batches[window]) >= self._batch_size:
+      with self._batch_size_estimator.record_time(self._batch_size):
+        yield windowed_value.WindowedValue(
+            self._batches[window], window.max_timestamp(), (window,))
+      del self._batches[window]
+      self._batch_size = self._batch_size_estimator.next_batch_size()
```

Comment:
<!--thread_id:cc_144435541_t; commit:85e7756335b0cf7501f62bd283b07fa9ea468a21; resolved:1-->
<!--section:context-quote-->
> **katsiapis** wrote:
> Is it easy (and does it make sense?) to have a window-specific batch size?

<!--section:body-->
I considered this, but the problem is that the set of windows may be unbounded (with new ones coming in over time) and we'd have to add logic to age them out over time. Also, I think it's reasonable to assume that the data in adjacent windows have similar characteristics, and so here we get the benefits of learning across windows.